lexicon/providers/aliyun.py

```diff
@@ -0,0 +1,217 @@
+
+    def __get_record(self, identifier):
+        response = self._request('DescribeDomainRecordInfo', query_params={'RecordId': identifier})
+
+        return {
+            'id': response['RecordId'],
+            'type': response['Type'],
+            'name': self._full_name(response['RR']),
+            'ttl': response['TTL'],
+            'content': response['Value']
+        }
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        resources = self._list_records(rtype, name, None)
+
+        for record in resources:
+            if (rtype == record['type']) \
+                and (self._relative_name(name) == self._relative_name(record['name'])) \
+                and (content == record['content']):
+                return True
+
+        if not identifier:
+            record = resources[0] if resources else None
+            identifier = record['id'] if record else None
+
+        if not identifier:
+            self._create_record(rtype, name, content)
+
+        LOGGER.debug('update_record: %s', identifier)
+
+        query_params = {'RecordId': identifier}
+
+        if rtype:
+            query_params['Type'] = rtype
+
+        if name:
+            query_params['RR'] = self._relative_name(name)
+
+        if content:
+            query_params['Value'] = content
+
+        query_params['TTL'] = self._get_lexicon_option('ttl')
+        self._request('UpdateDomainRecord', query_params=query_params)
+
+        return True
+
+    # Delete an existing record.
+    # If record does not exist, do nothing.
+    def _delete_record(self, identifier=None, rtype=None, name=None, content=None):
+        delete_resource_id = []
+        if not identifier:
+            resources = self._list_records(rtype, name, content)
+            delete_resource_id = [resource['id'] for resource in resources]
+        else:
+            delete_resource_id.append(identifier)
+
+        LOGGER.debug('delete_records: %s', delete_resource_id)
+
+        for resource_id in delete_resource_id:
+            self._request('DeleteDomainRecord', query_params={'RecordId': resource_id})
+
+        return True
+
+    def _request(self, action='GET', url='/', data=None, query_params=None):
+        # force set to the fixed api endpoint
+        url = ALIYUN_DNS_API_ENDPOINT
+
+        if query_params is None:
+            query_params = {}
+
+        query_params.update(self._build_default_query_params(action))
+        query_params.update(self._build_signature_parameters())
+
+        query_params.update({'Signature': self._calculate_signature('GET', query_params)})
+
+        response = requests.request('GET', url, params=query_params)
+
+        # TODO: add exception handle
+        return response.json()
+
+    # @param query_params should be {} and not none
+    def _calculate_signature(self, http_method, query_params):
+        access_secret = self._get_provider_option('access_secret')
```

Comment:
`access_secret` may be `None` if the user forgot to pass the flag. It is reasonable to raise an exception here in case of `None` value, and avoid to send requests that have no chance to be authenticated.